File: api/folder_move.go
package vaku

import (
	"context"
	"errors"
)

var (
	// ErrFolderMove when FolderMove fails.
	ErrFolderMove = errors.New("folder move")
)

// FolderMove moves data at a source folder to a destination folder. Source is deleted after copy.
func (c *Client) FolderMove(ctx context.Context, src, dst string) error {
	err := c.FolderCopy(ctx, src, dst)
	if err != nil {
		return newWrapErr("", ErrFolderMove, err)
	}

	err = c.FolderDelete(ctx, src)
	if err != nil {
		return newWrapErr("delete "+src, ErrFolderMove, err)
	}

	return nil
}

// FolderMoveAllVersions moves all versions of each secret at a source folder to a destination folder (KV v2 only). Source is deleted after copy.
func (c *Client) FolderMoveAllVersions(ctx context.Context, src, dst string) error {
	err := c.FolderCopyAllVersions(ctx, src, dst)
	if err != nil {
		return newWrapErr("", ErrFolderMove, err)
	}

	err = c.FolderDelete(ctx, src)
	if err != nil {
		return newWrapErr("delete "+src, ErrFolderMove, err)
	}

	return nil
}

File: src/api/folder_move.rs
//! `FolderMove` — folder_copy + folder_delete on source.

use crate::api::client::Client;
use crate::api::error::{Error, ErrorKind};

impl Client {
    /// Recursive move (copy + delete) from `src` to `dst`.
    pub async fn folder_move(&self, src: &str, dst: &str) -> Result<(), Error> {
        self.folder_copy(src, dst)
            .await
            .map_err(|e| Error::wrap("", ErrorKind::FolderMove, Some(Box::new(e))))?;
        self.folder_delete(src).await.map_err(|e| {
            Error::wrap(
                &format!("delete {src}"),
                ErrorKind::FolderMove,
                Some(Box::new(e)),
            )
        })
    }
}
